work version

Victoria9Bob opened the pull request:
Посмотри, пожалуйста, проект. Особенно ApiDefectTest, так и не получилось массив объектов создать так, чтобы не ругалось. 
CI подключу чуть позже. Спасибо

Victoria9Bob committed:
wor version

voropa commented:
.idea и .iml файлы должны быть в .gitignore, так как это настройки твоей среды разработки

Victoria9Bob committed:
Add steps

voropa commented on src/test/java/Adapters/BaseAdapter.java:
```diff
@@ -0,0 +1,48 @@
+public abstract class BaseAdapter {
+    final String BASE_URL = "https://api.qase.io/v1/";
+//    final String AUTRTH_TOKEN = "4f7faca87e882d98cce5f4234195ac3c24f22d51";
+    final String AUTRTH_TOKEN = "74d33400e63f102e391e5e1bc8cb9d365cd9d350";
```
Такие вещи как токен и логин/пароль лучше хранить и читать из .ocnfig файла, так как это давольно секьюрные данные и плохо, когда они лежат в коде в public репозитории на github

voropa replied:
Вот тут я показывал  пример, как это сделать
https://docs.google.com/presentation/d/18CuJJrbIL2JbnZvoHOvA9Qw2JV97TLEhZSxj6LuV_UE/edit#slide=id.ga668f9158c_0_29

Victoria9Bob replied:
Этим сегодня займусь

Victoria9Bob replied:
теперь читает через property reader

voropa commented on src/test/java/Elements/BaseElement.java:
```diff
@@ -0,0 +1,24 @@
+    public void scrollIntoView(WebElement element) {
+        ((JavascriptExecutor) driver).executeScript("arguments[0].scrollIntoView(true);", element);
+        try {
+            Thread.sleep(500);
```
Плохо использовать thread.sleep() и зачем он т тут вообще?

Victoria9Bob replied:
Убрала

voropa commented on src/test/java/Elements/Drobdown.java:
```diff
@@ -0,0 +1,25 @@
+
+@Log4j2
+public class Drobdown extends BaseElement {
+    final String drobdownLOC = "//*[text()='%s']/ancestor::div/div/div[contains(@class,'notranslate')]";
```
Не очень хорошее название переменных, точнее оно не соответсвует java code convention. Имена переменных должны быть в camelCase, т.е dropdownLoc, а вообще "Loc" не очень хорошее сокращение для locator. так что лучше не полениться и дописать слово locator полность)
Да, и droPdown, а не droBdown.

voropa commented on src/test/java/Elements/Drobdown.java:
```diff
@@ -0,0 +1,25 @@
+        WebElement optionToClick = driver.findElement(By.xpath(String.format(optionLOC, label, optionName)));
+        scrollIntoView(optionToClick);
+        optionToClick.click();
+        log.info(String.format("Select an option: %s", optionName));
```
Я как-то говорил, что логи нужно  писать перед  действием, а не после него, чтобы в случае падения теста было понятно на каком шаге он  упал

Victoria9Bob replied:
Переставила

voropa commented on src/test/java/Elements/EditSuiteElement.java:
```diff
@@ -0,0 +1,26 @@
+@Log4j2
+public class EditSuiteElement extends BaseElement {
+    final String textInputLOC = "//*[text()='%s']/ancestor::div/div/input";
+    final String drobdownLOC = "css-io3r9z-singleValue";
```
Выглядит как не очень надежный локатор...

Victoria9Bob replied:
Поправила

voropa commented on src/test/java/Elements/FormatedTextElement.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public String getSuiteName(String caseName) {
+        log.info(String.format("Take value from %s", label));
```
Вот тут уже нарушается принцип абстракции при использовании классов-декораторов. 
FormatedTextElement должен работать уметь работать с любым элементом, у которого схожий локатор и схожее поведение и наоборот, любой элемент на странице, у которого схожий локатор и поведение может быть обернут в FormatedTextElement.
Так что методы getValue(), getVisibleText(), getTitleText() отлично под это подпадают и теоретически у любого элемента они могут быть вызваны.
А вот методы getSuiteName() и getStepText() нарушают этот принцип, так как они применимы только для каких-то определенных suite и step элементов и вряд ли  применимы для любых других элементов, обернутых в FormatedTextElement.
Так что тебе нужно либо пересмотреть элементы и локаторы и либо сделать еще отдельные декораторы для suite и step элементов, либо адаптировать их таким образом, чтобы getSuiteName и getStepText получались через общие getValue() или getVisibleText() методы

Victoria9Bob replied:
Создала отдельные классы-обертки, где это было нужно

voropa commented on src/test/java/Elements/Input.java:
```diff
@@ -0,0 +1,28 @@
+        log.info(String.format("Setting %s into %s input", text, label));
+
+    }
+    @Step("Setting {text} into step input")
```
То же самое тут, нарушается абстрактность Input декоратора

Victoria9Bob replied:
исправила

voropa commented on src/test/java/Modals/CaseDetailsModal.java:
```diff
@@ -0,0 +1,86 @@
+    }
+    @Step("Get case details info")
+    public Case getCaseDetailsInfo(String caseName) {
+        Case aCase = new Case();
```
Что за странное название переменной? 
Лучше называт ьтакие вещи через result, например resultCase

Victoria9Bob replied:
сделала

voropa commented on src/test/java/Modals/CaseModal.java:
```diff
@@ -0,0 +1,91 @@
+    public CaseModal(WebDriver driver) {
+        super(driver);
+    }
+    @Step("Fill form.")
```
Весьма неинформативное описание step'a. Лучше добавить хотя бы название формы, которая заполняется

Victoria9Bob replied:
сделала

voropa commented on src/test/java/Modals/CaseModal.java:
```diff
@@ -0,0 +1,91 @@
+        return this;
+    }
+    @Step("Save Case.")
+    public RepositoryPage saveCaseButton() {
```
Названия методов должны быть глаголами, обозначающими какое-то действие, а в твоем случае это имена существительные. Так что должно быть, например, clickSaveCaseButton вместо saveCaseButton. 
ниже то же самое. И проверь во всех остальных page-классах

Victoria9Bob replied:
поменяла названия

voropa commented on src/test/java/Models/Case.java:
```diff
@@ -0,0 +1,72 @@
+                Objects.equals(type, aCase.type) && Objects.equals(layer, aCase.layer) && Objects.equals(isFlaky, aCase.isFlaky) && Objects.equals(milestone, aCase.milestone) &&
+                Objects.equals(behavior, aCase.behavior) && Objects.equals(automationStatus, aCase.automationStatus) && Objects.equals(pre_conditions, aCase.pre_conditions) &&
+                Objects.equals(post_conditions, aCase.post_conditions) && Objects.equals(action, aCase.action) && Objects.equals(inputData, aCase.inputData) &&
+                Objects.equals(expectedResult, aCase.expectedResult);
```
Методы equals and hasCode генерируются автоматически при помощи аннотации @Data
А для конструктора есть анотация @AllArgConstructor
Так что можно заметит эти 100 строчек кода на 2 аннотации)

Victoria9Bob replied:
сделано

voropa commented on src/test/java/Models/SelectTestCases.java:
```diff
@@ -0,0 +1,4 @@
+package Models;
```
пустой класс, удали его, если он нигде не используется

Victoria9Bob replied:
удалила

voropa commented on src/test/java/Models/Suite.java:
```diff
@@ -0,0 +1,38 @@
+    private String description;
+    private String preconditions;
+
+    public Suite(String suiteName, SuiteParentSuite parentSuite, String description, String preconditions) {
```
То же самое тут, можно заменить на @Data и @AllArgConstructor

Victoria9Bob replied:
заменила на аннотации и удалила лишние строчки кода

voropa commented on src/test/java/Pages/LoginPage.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @Step("Open page.")
+    @Link(BASE_URL)
```
C помощью аннотаций @Link в allure обычно прикрепляют ссылки на какой-нибудь дефект или задачу из project management системы, а не указывают url страницы, на  которой сейчас выполняется тест. 
Так что ты их тут неправильно используешь и они излишние.

Victoria9Bob replied:
удалила ссылки в аннотациях

voropa replied:
А вот именно тут не удалила)

Victoria9Bob replied:
готово

voropa commented on src/test/java/Pages/ProjectsPage.java:
```diff
@@ -0,0 +1,35 @@
+import org.openqa.selenium.WebDriver;
+
+public class ProjectsPage extends BasePage {
+    protected final static String ProjectsURL = "https://app.qase.io/projects";
```
константы должны быть написаны в UPPER_CASE

Victoria9Bob replied:
поправила модификаторы доступа и названия локаторов

voropa commented on src/test/java/Pages/RepositoryPage.java:
```diff
@@ -0,0 +1,98 @@
+    protected final static String repositoryURL = "https://app.qase.io/project/DEMO?view=1&suite=2";
+    final By CREATE_CASE = By.id("create-case-button");
+    final By CREATE_SUITE = By.id("create-suite-button");
+    private String EDIT_SUITE = "//span[contains(text(),'%s')]//following::button[2]";
```
Плохие локаторы с порядковыми номерами внутри типа button[2], лучше исправить

Victoria9Bob replied:
Убрала

voropa commented on src/test/java/Tests/CreateCaseTest.java:
```diff
@@ -0,0 +1,47 @@
+@Log4j2
+public class CreateCaseTest extends BaseTest{
+    @Test
+    public void createCaseWithAllData() throws InterruptedException {
```
Так  как в этом классе всего один тестовый метод, то можно для красоты вынести создание тестовых данных в BeforeMethod, а удаление тестовых данных в AfterMethod

Victoria9Bob replied:
Вынесла в классах CaseTest, SuiteTest, PlanTest

voropa commented on src/test/java/Tests/CreateSuiteTest.java:
```diff
@@ -0,0 +1,35 @@
+@Listeners(TestListener.class)
+public class CreateSuiteTest extends BaseTest {
+    @Test
+    public void createSuiteWithAllData() throws InterruptedException {
```
То же самое с before и after method

Victoria9Bob replied:
сделано

voropa commented on src/test/java/Tests/TestListener.java:
```diff
@@ -0,0 +1,69 @@
+        log.error("Test "+result.getName()+" failed.");
+//        this.driver = ((BaseTest)result.getInstance()).driver;
+//        File scrFile = ((TakesScreenshot)driver).getScreenshotAs(OutputType.FILE);
+//        File distfile = new File("/D:/Qase/target/"+result.getName()+".scr.jpg");
```
Почему тут закоментированный код? либо удали его, либо раскоментируй

Victoria9Bob replied:
сделано)

voropa commented on src/test/resources/allure.properties:
```diff
@@ -0,0 +1,3 @@
+allure.results.directory=target/allure-results
+allure.link.issue.pattern=https://instagram.com/{}
```
instagram?) он тут точно нужен?)

Victoria9Bob replied:
Удалила)

voropa commented on src/test/resources/regression.xml:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Почему только один suite.sml?
Я бы слелал как миниум 2:
в одном API тесты, в другом - UI тесты. Так как на реальных проектах чаще всего их запускают раздельно

Victoria9Bob replied:
Сделала 3 API и 2 на UI(smoke, regression). Добавила группы

voropa commented on src/test/resources/config.properties:
```diff
@@ -0,0 +1,3 @@
+qase.username = enter your username here
```
Вот, у тебя же даже есть config.properties файл. Вот в него нужно вынеcти все sensetive данные, типа username/password и token

Victoria9Bob replied:
теперь читает через property reader

voropa commented on pom.xml:
```diff
@@ -7,8 +7,89 @@
+    </build>
 
     <properties>
+        <suiteXmlFile>src/test/resources/regression.xml</suiteXmlFile>
```
Ты уверена, что у тебя запускается именно regression.xml по команде mvn test? По-поему в конфигурации maven-surefire-plugin выше пропущен атрибут suiteXmlFile, а должен быть
<img width="503" alt="image" src="https://user-images.githubusercontent.com/14195681/153769434-292e5c02-c0c4-41f4-86e2-4bf2200c7cd9.png">

Victoria9Bob replied:
Добавила

voropa commented on src/test/java/Tests/ApiDefectTest.java:
```diff
@@ -0,0 +1,86 @@
+                actual_result("A").
+                severity(3).
+                milestone_id(1).
+                custom_field(DefectCustomFields.builder().id("0").value("").build()).
```
Вот так работает
custom_field(Arrays.asList(DefectCustomFields.builder().id("0").value("").build())).
Плюс еще в Defect классе нужно поменять тип свойства custom_field на 
List<DefectCustomFields> custom_field;

По аналогии в строчке 61

Victoria9Bob replied:
Сработало, спасибо

Victoria9Bob replied:
Но сами методы не работают все равно)

Victoria9Bob replied:
Поправила тесты, они работают, но тест на создание дефекта возвращает только один параметр -id. Он его присваивает сам каждый раз новый и получается, что тест одноразовый) Нужно вручную выставлять ожидаемый id. Не знаю, как от этого уйти(

Victoria9Bob replied:
готово

voropa commented on src/test/java/Tests/ApiProjectTest.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Test
+    public void
+    createProjectPositiveTest() {
```
Этот тест падает у меня.
вероятно, проблема в том, что project code должен быть уникальным и система ругается:
    "errorFields": [
        {
            "field": "project",
            "error": "Project with the same code already exists."
        }
    ]

Victoria9Bob replied:
да, теперь генерит уникальный через Faker

voropa commented on src/test/java/Tests/ApiProjectTest.java:
```diff
@@ -0,0 +1,73 @@
+    @Test
+    public void
+    createProjectPositiveTest() {
+        String projectCode = "API";
```
А вот твой projectCode, который всегда одинаковый, вот и тест с ним может выполнится только один раз...
Так что нужно генерировать рандомный код, чтобы он прикаждом запуске был разный и тест всегда работал.
Библиотека java-faker тебе в помощь
https://github.com/DiUS/java-faker

Victoria9Bob replied:
Добавила библиотеку и прописала данные в тестовых классах через нее. Отдельно класс генератор не создавала, т.к. перегружать метод с разными вариантами мне не нужно

voropa commented on src/test/java/Tests/CreateCaseTest.java:
```diff
@@ -0,0 +1,47 @@
+
+@Listeners(TestListener.class)
+@Log4j2
+public class CreateCaseTest extends BaseTest{
```
У меня этот тест не проходит, все время падает. Предполагаю, что где-то нужно добавить ожидание....

Victoria9Bob commented:
> Файлы .idea и .iml должны быть в .gitignore, так как это настройки вашей среды разработки

Добавила файлы target .idea *.iml .allure в .gitignore

Victoria9Bob commented:
> .idea и .iml файлы должны быть в .gitignore, так как это настройки твоей среды разработки

Добавила

Victoria9Bob committed:
Fixed:
Steps,locators, annotations, links, before & after methods in tests, allure properties, custom_field class.

Victoria9Bob committed:
Fixed:
test names,log,variable names

Victoria9Bob committed:
Add Java Faker

Victoria9Bob committed:
Add xml smoke, regression,api. Add groups

Victoria9Bob committed:
Fixed wrapper classes

Victoria9Bob committed:
Fixed wrapper classes(Input)

Victoria9Bob committed:
Edit locators

voropa commented on src/test/java/Pages/LoginPage.java:
```diff
@@ -0,0 +1,80 @@
+        return this;
+    }
+
+    @Step("Open login page.")
```
Так у тебя же уже есть метод open() в этом классе. зачем тебюе еще один метод openLoginPage()??

Victoria9Bob replied:
Уже удалила) Потому что у меня для логина и создания пользователя url разные и чтобы не делать лишних действий и искать нужные страницы, сразу по url перехожу.

Victoria9Bob replied:
Ты вроде говорил, что так можно

voropa replied:
Очень вряд ли. Если используется loadablePage паттерн, то метод open() в каждой странице должен быть определен однозначно...

Victoria9Bob replied:
поправила

voropa commented on src/test/java/Pages/RepositoryPage.java:
```diff
@@ -0,0 +1,96 @@
+
+@Log4j2
+public class RepositoryPage extends BasePage {
+    protected final static String REPOSITORY_URL = "https://app.qase.io/project/DEMO?view=1&suite=2";
```
Думаю, что этот URL не сработает в 99% случаев, так как в нем указан код конкретного проекта.
Для таких страниц, которые зависят от какого-то обьекта и в URL имеют параметры этого обьекта, нет смысла делать метод open() через driver.get(URL). т.к. им нельзя воспользоваться в 99% случаев

Victoria9Bob replied:
Так мне и нужен конкретный проект, потому что в противном случае мне нужно создавать новый проект перед каждым тестом, что увеличивает время каждого теста. Поэтому я использую тестовый проект, а создание проектов проверяю отдельными тестами через api.

voropa replied:
Да, но конкретный проект - это часть  тестовых данных, а тестовые данные должны быть в тестах, а не захардкоданы в пэйджах. Пэйджы - это абстрактные классы, которые должны работать с любыми данными...
В таком случае тебе нужно хранить название тестового проекта где-то в тестах и открывать его каким-то методом типа homePage.openProject(testProjectName)

voropa replied:
Т.е. у тебя должна быть страница,  н которую ты попадаешь после авторизации, и уже на ней ты кликаешь на нужный проект. Тогда открывается RepositoryPage с проектом
<img width="1177" alt="image" src="https://user-images.githubusercontent.com/14195681/154833160-d97df0f4-8654-4033-82aa-16f4e441f236.png">

loginPage.login();
//now you're at homePage
homePage.openProject(projectName);
//now you're at project repositoryPage
repositoryPage.doSomeAtcions()......

Victoria9Bob replied:
Я понимаю, о чем ты говоришь. А можно оставить переход по url, но id проекта вынести отдельно в base класс и переопределяя метод open() как бы собирать нужный url, таким образом, если мне нужно зайти в другой проект, я меняю только одну переменную?Показала отдельным комитом.

voropa replied:
Да, так теретически тоже можно, это будет хотя бы правильно с точки зрения архитектуры. Метод open() должен получать на вход id проекта и формировать динамический URL.
Но на будущее правильнее делать так, как я писал  выше, потому что selenium тесты должны эмулировать действия пользователя, а пользователь вряд ли будет открывать проект через адреснуб строку, он скорее будет кликать на иконку на UI)

Соответсвенно переделай метод open() во всех классах-страницах, где в URL есть id проекта

Victoria9Bob replied:
Т.е. в каждом тесте все равно идет логин и открывается страница проектов, просто вместо действия кликнуть на проект(DEMO), я открываю его url, а не по локатору, со страницей test plans вместо кликнуть на проект, кликнуть на тест план - сразу открываю тест план))

Victoria9Bob replied:
поправила

voropa commented on src/test/java/Pages/RepositoryPage.java:
```diff
@@ -0,0 +1,96 @@
+        driver.get(REPOSITORY_URL);
+        return this;
+    }
+    //span[contains(text(),'%s')]//ancestor::h3//i[contains(@class,'pencil')]
```
удали закоментированный код

Victoria9Bob replied:
удалила

voropa commented on src/test/java/Pages/TestPlansPage.java:
```diff
@@ -0,0 +1,77 @@
+
+@Log4j2
+public class TestPlansPage extends BasePage {
+    protected final static String TEST_PLANS_URL = "https://app.qase.io/plan/DEMO";
```
то же самое тут, у тебя TestPlansPage привязывается к проекту, который ты открываешь, так что нельзя попасть на эту страницу по URL

Victoria9Bob replied:
поправила

voropa commented on src/test/java/Tests/CreateCaseTest.java:
```diff
@@ -0,0 +1,58 @@
+                expectedResult(faker.harryPotter().character()).
+                build();
+        boolean isloggedIn = loginPage.openLoginPage().login(EMAIL, PASSWORD).isPageOpened();
+        Assert.assertTrue(isloggedIn);
```
В before/After методах не должно быть проверок, проверки должны быть только  в тестах

Victoria9Bob replied:
Перекинула в тесты

voropa commented on src/test/java/Tests/CreateSuiteTest.java:
```diff
@@ -0,0 +1,44 @@
+                build();
+
+        boolean isloggedIn = loginPage.openLoginPage().login(EMAIL, PASSWORD).isPageOpened();
+        Assert.assertTrue(isloggedIn);
```
То же самое

Victoria9Bob replied:
сделала

voropa commented on src/test/java/Elements/EditSuiteElement.java:
```diff
@@ -0,0 +1,23 @@
+        super(driver, label);
+    }
+
+    public String getDropdownSuiteValue() {
```
Какой смысл от этого метода, если он не зависит ни от label'a в этом классе и сам локатор не параметризированный. По факту этот локатор ищет любой dropdown на любой странице, так что думаю, что ему не место в этом классе-обертке

Victoria9Bob replied:
Убрала

voropa commented on .gitignore:
```diff
@@ -1,3 +1,6 @@
 # Default ignored files
 /shelf/
-/workspace.xml
+target
```
Все эти фалы все еще попадают в пулл реквест.
Вероятно не совсем правильно указаны пути в .gitignore
Должно быть как то так
/.idea
/target
/.allure
/.allure-results
*.iml

Victoria9Bob replied:
поправила

voropa reviewed (changes requested):
1. Прожми кнопку resolve для тех комантариев, которые ты уже исправила, чтобы они не висели открытыми
2. Пересмотри оставшиеся незакрытые коментарии + несколько новых
3. ОТДЕЛЬНЫМ КОМИТОМ в самом конце отформатируй весь проект, чтобы убрать лишние пробелы, отступы и т.д 
В IDEA это можно сделать автоматически:
<img width="464" alt="image" src="https://user-images.githubusercontent.com/14195681/154799016-e95c85a4-a084-4799-b5be-89f9ed64eb83.png">

voropa commented:
Еще общее замечание: названия пакетов (Tests, Pages, Enums, ...) в java должны быть с маленькой буквы

Victoria9Bob committed:
Add JenkinsFile

Victoria9Bob committed:
Add PropertyReader

Victoria9Bob committed:
Updated configfile,changed url and moved the project ID separately

Victoria9Bob committed:
Fixed: beforeMethod in tests, wrapper class

Victoria9Bob committed:
Сhanged package names

voropa commented on src/test/resources/config.properties:
```diff
@@ -1,7 +1,6 @@
-qase.test_plans_url = https://app.qase.io/plan/DEMO
\ No newline at end of file
+qase.base_url = https://app.qase.io
+qase.repository_uri = /project
```
Эти пути незачем хранить в конфиг файле, они могут лежать прямо в page классе. В конфиг файле у тебя должен быть base_url и project_id.
А в page классах уже будет формироваться динамический URL, на основании этих параметров

Victoria9Bob committed:
Fixed defects tests

Victoria9Bob committed:
Update defect test

Victoria9Bob committed:
Reformated code

voropa commented:
Please resolve merge conflicts below

Victoria9Bob committed:
Update JenkinsFile

voropa commented on src/test/resources/config.properties:
```diff
@@ -0,0 +1,6 @@
+qase.email=bob123@mailinator.com
```
Все эти данные нужно убрать из property файла, перед тем, как мержить в master

Victoria9Bob replied:
заменила все данные в этом файле

voropa commented on src/test/java/tests/CreateSuiteTest.java:
```diff
@@ -0,0 +1,46 @@
+    @AfterMethod(groups = "Regression")
+    public void setOut() {
+        repositoryPage.deleteTestSuite(suiteName);
+//        Assert
```
Лишний коментарий

Victoria9Bob replied:
изменила

voropa commented on src/test/java/tests/ApiDefectTest.java:
```diff
@@ -0,0 +1,73 @@
+package tests;//package Tests;
```
Так же лишнйи коментарий

Victoria9Bob replied:
убрала

voropa commented on src/test/java/pages/BasePage.java:
```diff
@@ -0,0 +1,40 @@
+
+public abstract class BasePage {
+    protected final static String BASE_URL = System.getenv().getOrDefault("BASE_URL", PropertyReader.getProperty("qase.base_url"));
+    protected final static String PROJECT_ID = "/DEMO";
```
pROJECT_ID - это переменная, относящаяся к тестовым данным, а не к странице,  поэтому должна хранится скорее в BaseTest и передаваться в page в качестве входного аргумента в методах open()

Victoria9Bob replied:
сделала

voropa commented on src/test/java/models/Defect.java:
```diff
@@ -0,0 +1,18 @@
+    String title;
+    String actual_result;
+    int severity;
+//    int milestone_id;
```
закоментированый код

Victoria9Bob replied:
убрала

Commit:
added README.md

Victoria9Bob committed:
Update uri

Victoria9Bob committed:
Merge remote-tracking branch 'origin/WorkBranch' into WorkBranch
